### ISSUE TITLE
[Enhancement] Ai Editor 

### ISSUE THREAD

**Issue by Portals** [UnderwaterApps/overlap2d#148]
Hi! 

I'm in the process of learning gdx-AI, and I'm roughly halfway through. I would like to create a graphical AI editor. I have some ideas in my head, but I need to write them down first. Is there any interest for this?

**Comment by azakhary** [comment]
You mean to make it as a plugin? I am all for it as you can imagine :) But I am also curious how will that look.. exactly? Any thoughts on the flow? even draft-ish?

**Comment by Portals** [comment]
Yeah, like a plugin! I'm going to look into it later tonight when I get of my summer job and then create a draft!

**Comment by Portals** [comment]
This is just a quick summary of what I had in mind. More to come later! 

The most important part of gdx-AI is the (behavior tree)[https://github.com/libgdx/gdx-ai/wiki/Behavior-Trees]. WIth it you can control the steering of a NPC, you can control what it should do, and if it should send any messages. I would like to create this behavior tree graphical like a tree of nodes. Where you easily can create a new branch and so on. It's quite complex, there's going to be different kinds of branches / leafs and so on. 

The only thing I'm still not sure about is how I would connect the nodes to code in a nice way. Got to think about it a bit more. After the behavior tree is done, I can implement steering, messages and pathfinding(Pathfinding is going to be tough). 

I'm going to do a mindmap to connect everything together, and I need to read more into the Overlap2D source code so I know easier how to create the plugin, and I need to finish learning about gdx-AI. I will try to find time to work on this, but this next week, I'm going to be working my last week at my summer job, and I will mostly plan on my free time / learn about gdx-AI. After that I'll have more time on my hands.

This is just a quick draft, sorry if it's confusing. :)

**Comment by Portals** [comment]
Note that the gdx-AI [https://github.com/libgdx/gdx-ai/wiki/State-Machine] (state machine) also can be implemented! Both has it's strenght and weaknesses. Note that I'm not a gdx-AI expert, and still are learning!

**Comment by Portals** [comment]
TL;DR A graphical tree of nodes which is a hybrid between state machine and behavior tree.

**Comment by azakhary** [comment]
Have see checked this? http://forum.unity3d.com/threads/behavior-designer-behavior-trees-for-everyone.227497/

**Comment by Portals** [comment]
No, haven't had time to research yet. Just wanted to throw the idea out there asap.

**Comment by azakhary** [comment]
All I am saying, that's a good place to start inspiring, if Unity has it, we should have it too :-P and a better one.

**Comment by Portals** [comment]
I'll try to research as much as I can, though as I said. This week will be my last week in work! :)

**Comment by gj12** [comment]
This would be a great enhancement. An editor for AI/NPC behavior can save a lot of game-development time.

**Comment by Portals** [comment]
Are you planning to make the plugin system 100% independent from the main project? So you're able to add components and so on without having to touch the main project?

**Comment by Portals** [comment]
Also, I think I know how I should do this, will start work this weekend!

**Comment by Portals** [comment]
Question is if I should have a seperate project, has a plugin or if I should program directly into the main project.

**Comment by azakhary** [comment]
it is independent. plugin is totally external. But because system is new, it may have bugs, and you will have to fix it on main project to have something in your plugin. bu other then that they are separate.

**Comment by azakhary** [comment]
make external project, and inspire from example plugin I made which is called "performance"  I'll try to make short guide for that today.

My plugin is in the repo, but your plugin could be on your own repo. just create empty project, see how interfaces and sources are made for "performance" plugin.

In order t test things quickly, you can configure your plugin gralde to buuld jar in specific location (to plugins folder of your overlap2d) so every time you build, you can run o2d and see result.

**Comment by Portals** [comment]
Might be a stupid question. But how will I add what's necessary in runtime? Do I have to create two plugins, one for editor, and one for runtime?

**Comment by azakhary** [comment]
yes generally if your plugin saves some additional data, you might want to have a "runtime" piece that will read it. At least that's how I imagined it :D Btw, please do ask me questions on your process a lot, the plugin system is new, and whatever problems you have I am here to help. (or maybe add API to it)

**Comment by Portals** [comment]
I will close this issue for now, will come back to it later though.

**Comment by azakhary** [comment]
How is it going by the way, did you make anythin plugin-ish work?

**Comment by Portals** [comment]
I have not created any editor, I did test VisWindow a bit and played around that a bit. Also created this mockup a while ago in powerpoint. 

http://imgur.com/3AxWOpw

A way to combine states and behavior tree. Not 100% sure how the scripts part or events part should work though. I will come back to this later, and when I do I should have something to show. Expect something in a month or 2. Sorry for delaying this...

**Comment by azakhary** [comment]
I'll try to add some thoughts to this, I happened to use gdx-ai lately.

**Comment by Portals** [comment]
I created this repository: https://github.com/Portals/BlueAiEditor 

Will try to keep this updated.
